feat(manager/cargo): support git dependencies

okkero opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

This PR reintroduces Cargo git dependencies, reverted in #36412. The [original PR](https://github.com/renovatebot/renovate/pull/36289) adding support for git dependencies also introduced two regressions. This PR addresses both:

- #36397 is addressed by not using `cargo` versioning by default for git dependencies, instead making a best effort decision between `semver` and `loose` based on the current value's shape.
- #36408 is addressed by not trying to read `currentValue` if it is not defined, effectively skipping git dependencies without a specified ref, leaving it up to the lock file maintenance schedule to update the lock file.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

- Closes #26531 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

mkniewallner committed:
feat(manager/cargo): support git dependencies

mkniewallner committed:
feat(manager/cargo): add git sources to `supportedDatasources`

mkniewallner committed:
fix(manager/cargo): only warn for non-locked on crates

mkniewallner committed:
Merge branch 'main' of github.com:mkniewallner/renovate into feat/support-git-dependencies-cargo

mkniewallner committed:
test(manager/cargo): add datasource for deps

okkero committed:
Merge remote-tracking branch 'origin/main' into feat/support-git-dependencies-cargo

okkero committed:
Merge remote-tracking branch 'origin/main' into feat/support-git-dependencies-cargo

okkero committed:
Default to https for registryUrl git dependency protocol is ssh

okkero committed:
Cargo: if dep.currentValue is unset when getting locked version, default to '*' instead of crashing.

This is necessary in cases where you have a git dependency without a
specified ref (no `rev`, `tag` or `branch`)

okkero committed:
Cargo: Never use `cargo` versioning for git dependencies

`cargo` versioning never makes sense for git dependencies, since the ref
always has to be an exact match.

If the current value looks like semver, we use `semver` versioning.
Otherwise we use `loose` versioning.

okkero committed:
Explicitly skip getting lockedVersion if currentValue is not set

okkero committed:
Test that git dependencies do not extract locked version

RahulGautamSingh commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -120,14 +121,24 @@ async function updateArtifactsImpl(
+        (dep) => !dep.lockedVersion && dep.datasource === CrateDatasource.id,
+      );
+      // Non-crate dependencies (like git ones) do not have locked versions.
+      // For crate dependencies, not having a locked version is not expected.
```
```suggestion
      // For crate dependencies, a locked version is expected.
```

RahulGautamSingh commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -120,14 +121,24 @@ async function updateArtifactsImpl(
+      // In both situations, perform a regular workspace lockfile update.
+      if (nonCrateDep || crateDepWithoutLockedVersion) {
+        if (crateDepWithoutLockedVersion) {
+          // Only warn when a crate dependency has no locked version, as this is
+          // not an expected situation.
```
```suggestion
          // Only warn when a crate dependency has no locked version
```

okkero commented on lib/modules/manager/util.ts:
```diff
@@ -21,7 +21,9 @@ export function applyGitSource(
       const { protocol, source, full_name } = parseGitUrl(git);
-      dep.registryUrls = [`${protocol}://${source}`];
+      dep.registryUrls = [
+        `${protocol === 'ssh' ? 'https' : protocol}://${source}`,
```
My thought process was this:

Cargo (maybe others as well) allows you to use ssh protocol for git dependencies (see https://doc.rust-lang.org/cargo/appendix/git-authentication.html#ssh-authentication). However, if you do so, that's still not the correct protocol for the registry URL (that would be `https`). Running on my test repository failed without this, so I thought coercing `ssh` to `https` was a sane default in this case.

However, I'm really unsure about the best approach to this, and I'm open to suggestions. Maybe something should be done at the cargo manager level instead...?

RahulGautamSingh replied:
This helper function can be useful here:
https://github.com/renovatebot/renovate/blob/03542cb78972d3b19efc25a6c89d83437dd1ccaf/lib/util/git/url.ts#L11

Btw, if you have difficulty understanding something, you can check the `poetry` and `vendir` managers. We extract git dependencies there.

okkero committed:
Let crate datasource handle default versioning for cargo

Co-authored-by: Michael Kriese <michael.kriese@gmx.de>

okkero committed:
Set type and initial value for lockedVersion

okkero committed:
Apply suggested comment changes

Co-authored-by: RahulGautamSingh <rahultesnik@gmail.com>

okkero committed:
Use consistent naming for import of `semverVersioning`

RahulGautamSingh commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -120,14 +121,23 @@ async function updateArtifactsImpl(
-          { dependency: missingDep.depName },
-          'Missing locked version for dependency',
-        );
+      const nonCrateDep = updatedDeps.find(
```
```suggestion
      const nonCrateDep = updatedDeps.some(
```
Since, we only need to value not the object

RahulGautamSingh reviewed (commented):
LGTM, except the comments and the protocol change situation

RahulGautamSingh commented on lib/modules/manager/cargo/schema.ts:
```diff
@@ -56,6 +60,21 @@ const CargoDep = z.union([
+            dep.versioning = semanticVersioning.id;
+          } else {
+            dep.versioning = looseVersioning.id;
+          }
```
You can remove this if-check and also remove the ` versioning: cargoVersioning.id,` from the `defaultConfig` in the `manager/cargo/index.ts` file.

Now, the versioning for each dependency will be its datasource's default versioning.

You can try this in your test repos, and if all works fine and the regressions are fixed. We can use this approach.

okkero commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -120,14 +121,23 @@ async function updateArtifactsImpl(
+      // Non-crate dependencies (like git ones) do not have locked versions.
+      // For crate dependencies, a locked version is expected.
+      // In both situations, perform a regular workspace lockfile update.
+      if (nonCrateDep || crateDepWithoutLockedVersion) {
```
connected to above suggestion

```suggestion
      if (hasNonCrateDep || crateDepWithoutLockedVersion) {
```

okkero committed:
Apply suggestions from code review

Co-authored-by: RahulGautamSingh <rahultesnik@gmail.com>

okkero committed:
Merge remote-tracking branch 'origin/main' into feat/support-git-dependencies-cargo

okkero committed:
Use getHttpUrl instead of manually replacing protocol in applyGitSource

okkero committed:
Fix lint errors

okkero committed:
Remove versioning override from cargo manager

okkero committed:
Update cargo snapshot to remove versioning requirement

RahulGautamSingh commented on lib/modules/manager/cargo/__snapshots__/extract.spec.ts.snap:
```diff
@@ -282,23 +282,63 @@ exports[`modules/manager/cargo/extract > extractPackageFile() > extracts multipl
   {
-    "currentValue": "0.1.0",
-    "datasource": "crate",
+    "currentValue": undefined,
+    "datasource": "git-refs",
     "depName": "git_dep_with_version",
     "depType": "dependencies",
     "managerData": {
       "nestedVersion": true,
     },
-    "skipReason": "git-dependency",
```
The version was detected before, but now it is unspecified? 🤔

rarkins replied:
Good catch. Why is `currentValue` now undefined?

okkero replied:
I think it's this: https://github.com/renovatebot/renovate/pull/36444/commits/8bbf20350551e2279a3390b26f080eabf224a458#diff-09657ce916b4a05cf4f9d8b64a7c5fc2eb4e4d6078cf7c45583656dc230899c1R65

@mkniewallner Can you confirm?

RahulGautamSingh replied:
I don't think any change is needed now after [this comment ](https://github.com/renovatebot/renovate/pull/36444#discussion_r2309783005)

viceice commented on lib/modules/manager/cargo/extract.ts:
```diff
@@ -284,10 +284,15 @@ export async function extractPackageFile(
-      );
+
+      let lockedVersion: string | null = null;
+      if (dep.currentValue) {
```
why this? can't it have a locked version without a pinned current value?

viceice commented on lib/modules/manager/cargo/index.ts:
```diff
@@ -17,7 +20,12 @@ export const categories: Category[] = ['rust'];
 export const defaultConfig = {
   commitMessageTopic: 'Rust crate {{depName}}',
   managerFilePatterns: ['/(^|/)Cargo\\.toml$/'],
-  versioning: cargoVersioning.id,
```
ensure it the default on crate datasource

okkero committed:
Merge remote-tracking branch 'origin/main' into feat/support-git-dependencies-cargo

okkero committed:
Remove unused import

okkero committed:
Merge remote-tracking branch 'origin/main' into feat/support-git-dependencies-cargo

okkero committed:
Add test case for cargo path dependencies without version

okkero commented on lib/modules/manager/cargo/__fixtures__/Cargo.1.toml:
```diff
@@ -20,6 +20,9 @@ pcap-sys = { version = "=0.1", path = "pcap-sys" }
 pnet = { version = "0.21.0", optional = true, default-features = false}
 git_dep_with_version = { version = "0.1.0", git = "https://github.com/foo/bar" }
```
No, this should still be skipped I think, since a git dependency doesn't make sense without a rev, tag or branch. Specifying `version` for a git dependency is more of a supplementary sanity check.

RahulGautamSingh replied:
Your reasoning sounds good. I think a comment/note somewhere will do the job if not then we can keep it as `git-dep-with-no-tag`.

okkero replied:
I'm not sure if this is what you meant, but I added a comment in the test fixture, explaining the behavior.

RahulGautamSingh reviewed (commented):
Added comment

okkero committed:
Add comment about git dependencies and version

Add a comment explaining the behavior of git dependencies without
rev/tag/branch and the version field

okkero commented on lib/modules/manager/cargo/__fixtures__/Cargo.1.toml:
```diff
@@ -18,8 +18,12 @@ libc = "=0.2.43"
+# If not, they should be skipped, even if version is defined.
 git_dep_with_version = { version = "0.1.0", git = "https://github.com/foo/bar" }
 git_dep = { git = "https://github.com/foo/bar" }
+
```
Do you disagree with the grouping? It was to make it clear that the comment applies only to those two deps.
